var/spack/repos/builtin/packages/py-numba/package.py

```diff
@@ -12,32 +12,38 @@ class PyNumba(PythonPackage):
     homepage = "https://numba.pydata.org/"
     pypi = "numba/numba-0.35.0.tar.gz"
 
+    version('0.55.1', sha256='03e9069a2666d1c84f93b00dbd716fb8fedde8bb2c6efafa2f04842a46442ea3')
     version('0.54.0', sha256='bad6bd98ab2e41c34aa9c80b8d9737e07d92a53df4f74d3ada1458b0b516ccff')
     version('0.51.1', sha256='1e765b1a41535684bf3b0465c1d0a24dcbbff6af325270c8f4dad924c0940160')
     version('0.50.1', sha256='89e81b51b880f9b18c82b7095beaccc6856fcf84ba29c4f0ced42e4e5748a3a7')
     version('0.48.0', sha256='9d21bc77e67006b5723052840c88cc59248e079a907cc68f1a1a264e1eaba017')
     version('0.40.1', sha256='52d046c13bcf0de79dbfb936874b7228f141b9b8e3447cc35855e9ad3e12aa33')
     version('0.35.0', sha256='11564937757605bee590c5758c73cfe9fd6d569726b56d970316a6228971ecc3')
 
-    depends_on('python@3.7:3.9', type=('build', 'run'), when='@0.54.0:')
-    depends_on('python@3.6:', type=('build', 'run'), when='@0.48.0:')
+    depends_on('python@3.7:3.10', type=('build', 'run'), when='@0.55.0:')
+    depends_on('python@3.7:3.9', type=('build', 'run'), when='@0.54')
+    depends_on('python@3.6:3.9', type=('build', 'run'), when='@0.53')
+    depends_on('python@3.6:3.8', type=('build', 'run'), when='@0.52')
+    depends_on('python@3.6:', type=('build', 'run'), when='@0.48:0.51')
```

Comment:
I wonder if we should guard this with 3.7/3.8 as an upper bound anyway. Once Python 3.11 is released Spack will automatically concretize to this version in order to use Python 3.11. Same for numpy and llvmlite below.

Reply:
I understand what you mean and in a way its the same as what I was thinking when encountering the issue above. The question still is: Set it to what?. For example as far as I can tell `py-numba@0.51.1` works fine with `python@3.9.10` (`spack install --test=root` and a small example taken taken from the webpage both work fine). The same goes for `py-numpy` (see comment above).

Reply:
I would just set it to the max version of the closest release (probably the closest newer release). Authors will unfortunately pin the max version of dependencies even if they technically work with newer versions. This is usually because they can't actually test those versions (they weren't released at the time) and they want to avoid a situation in which a newer version doesn't work.

Reply:
I just checked the different `setup.py` files for the `py-llvmlite` dependency and found this:
```
depends_on('py-llvmlite@0.31:', type=('build', 'run'), when='@0.47')
depends_on('py-llvmlite@0.30:', type=('build', 'run'), when='@0.46')
depends_on('py-llvmlite@0.29:', type=('build', 'run'), when='@0.44,0.45')
depends_on('py-llvmlite@0.28:', type=('build', 'run'), when='@0.43')
depends_on('py-llvmlite@0.27:', type=('build', 'run'), when='@0.42')
depends_on('py-llvmlite@0.26:', type=('build', 'run'), when='@0.41')
depends_on('py-llvmlite@0.25:', type=('build', 'run'), when='@0.40')
```
Starting with `py-numba@0.48`, `py-llvmlite` is pinned down. So I guess pinning the `py-llvmlite` versions down for the older versions as well should be fine.

For `python` and `py-numpy`, if I understand you correctly, you mean setting upper limits to:
``
depends_on('python@3.6:3.8', type=('build', 'run'), when='@0.48:0.51')

depends_on('py-numpy@1.15:1.20', type=('build', 'run'), when='@0.48:0.53')
```
although we would then "lose" a working combination, right?